### DISCUSSION TITLE
Has anyone deployed Ts.ED to vercel?

### DISCUSSION THREAD

**Discussion by rizchaerul** [TypedProject/ts-express-decorators#1636] [Q&A] [answered]
Hi, I tried deploying Ts.ED to vercel (i know it's not a good idea since it's serverless). And I got this error:

![image](https://user-images.githubusercontent.com/72289631/142010425-eb74d6a0-1adb-49a5-8afb-8a0da2ba4394.png)

What I did is:
1. tsed init .
2. add vercel json
3. change index.ts to main.ts
4. add to github
5. add to vercel

here's my vercel.json

```
{
  "builds": [
    {
      "src": "src/main.ts",
      "use": "@vercel/node"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "src/main.ts"
    }
  ]
}
```

Any help would be appreciated, thanks!

**Comment by Romakita** [comment, marked as answer]
Hello @rizchaerul 

The message said, the main.js doesn't export a function. Probably because, Vercel handle request like this:
```typescript
module.exports = (req, res) => {
  const { name = 'World' } = req.query;
  res.send(`Hello ${name}!`);
};
```

So you have to do the same thing with Ts.ED. I haven't time give you an example. But I'll try to you that ASAP :)

See you
Romain

**Reply by rizchaerul** [reply]
Thanks for answering!

But when i tried NestJS, It can run in vercel with no error, the main.ts is similar to Ts.ED (or maybe it's very different under the hood?)
![image](https://user-images.githubusercontent.com/72289631/142016060-1a1fcf8c-9eca-40f6-8b86-5abef00b07da.png)

**Reply by Romakita** [reply]
You’ve right it’s not really different to bootstrap the application

**Reply by Romakita** [reply]
Hello @rizchaerul 
Vercel make me crazy. The first point is, I don't understand how Nestjs work correctly with Vercel without providing a default exported function. Vercel documentation doesn't mention anything about that. The only official way to expose an express app like is documented here:
https://vercel.com/guides/using-express-with-vercel

So I started a basic application by following the official documentation:

https://github.com/Romakita/tsed-vercel

But the application behavior is not the same as expected when I run the app in local. There are some changes performed by vercel, and I don't understand why and how it's possible (because vercel doesn't provide enough logs)

So right now, the app start but the controllers and services isn't mounted correctly... this point is pretty strange. 

I'll keep my investigation...
See you
Romain

**Reply by rizchaerul** [reply]
Thank you for your answer, I appreciate it.

I see, that's really weird. Recently I just found out about vercel cli. Using this tool I can run the app with @vercel/node runtime locally. And TS.ED run perfectly fine without any issues (https://github.com/rizchaerul/tsed-vercel-cli.git just run `vercel dev`).

![image](https://user-images.githubusercontent.com/72289631/143007101-ae44ecc9-69b3-40d3-beb5-2c0e72bd1e4e.png)

But when I deploy the project to vercel server, the error comes up. So if it can run locally using vercel's node runtime, but can't run on the server, maybe it's not TS.ED problem after all? But rather a bug in vercel maybe?

**Reply by Romakita** [reply]
Hello @rizchaerul,

I worked to fix some issue on how the tsed modules are packaged. Now I have a working example :)
I'll finish to write the repository example.

But it's a good news. Now Ts.ED works with vercel (and any Serverless platform) !

See you
Romain

**Reply by Romakita** [reply]
Hello @rizchaerul 
Here is the repository: https://github.com/tsedio/tsed-example-vercel
It works :) https://tsed-vercel.vercel.app/  !!!

See you
Romain

**Reply by rizchaerul** [reply]
Woah, thank you so much! Will try this later.
